applications/tari_watcher/src/port.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2024 The Tari Project
+// SPDX-License-Identifier: BSD-3-Clause
+use std::{
+    io,
+    net::{IpAddr, SocketAddr},
+};
+
+use tokio::net::TcpListener;
+
+use crate::config::InstanceType;
+
+#[derive(Clone)]
+pub struct PortAllocator {
+    pub validator: ValidatorPorts,
+    pub wallet: MinotariPorts,
+}
+
+impl PortAllocator {
+    pub fn new() -> Self {
+        Self {
+            validator: ValidatorPorts::new(),
+            wallet: MinotariPorts::new(),
+        }
+    }
+
+    pub async fn open_at(&mut self, instance: InstanceType, name: &'static str) -> io::Result<u16> {
+        let addr = TcpListener::bind(SocketAddr::from(([127, 0, 0, 1], 0))).await?;
+        let port = addr.local_addr()?.port();
```

Comment:
Ports should be configurable by the user. The use cases for the watcher and the swarm are very different.
The swarm is for development ("As a developer, I want to start up a DAN network so that I can test my changes"), the watcher should be a tool that validator node operators can use ("As a node operator, I want to ensure my VN is registered and running, so that I can earn fees").

The main watcher feature is taking care of validator registration for you.

RE the ports, an operator may want to specify and expose ports as needed, so should be left up to them how to configure it. They can configure it by editing the config file. We could think about adding a feature down the line to allow the watcher to configure things, but for now we can keep this dead simple and have the validator node process use the configured ports in it's config - the watcher simply has to start the process and monitor it.

Reply:
Thanks for the clarification. Would you have me add the additional fields to the config for the validator node: `vn_jrpc` and `vn_http`? Since we don't need to monitor the wallet.

Reply:
These ports are already configurable in the VN config file, so I think we can leave it to the VN code.

Reply:
For now the watcher can simply start a validator node e.g `/path/to/tari_validator_node -c path/to/config.toml`. It does not need to know the JSON-RPC or web UI ports.

Reply:
I will remove the env stuff, I was unsure where it would get the port data from. We can work under the assumption now that it already knows it then, as you mentioned earlier.